samples/replicated/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala

```diff
@@ -0,0 +1,279 @@
+
+    def isCheckedOut: Boolean =
+      checkoutDate.isDefined
+
+    def hasItem(itemId: String): Boolean =
+      items.contains(itemId)
+
+    def isEmpty: Boolean =
+      items.isEmpty
+
+    def updateItem(itemId: String, quantity: Int): State = {
+      quantity match {
+        case 0 => copy(items = items - itemId)
+        case _ => copy(items = items + (itemId -> quantity))
+      }
+    }
+
+    def removeItem(itemId: String): State =
+      copy(items = items - itemId)
+
+    def checkout(now: Instant): State =
+      copy(checkoutDate = Some(now))
+
+    def toSummary: Summary =
+      Summary(items, isCheckedOut)
+  }
+  object State {
+    val empty =
+      State(items = Map.empty, checkoutDate = None)
+  }
+
+  /**
+   * This interface defines all the commands (messages) that the ShoppingCart actor supports.
+   */
+  sealed trait Command extends CborSerializable
+
+  /**
+   * A command to add an item to the cart.
+   *
+   * It replies with `StatusReply[Summary]`, which is sent back to the caller when
+   * all the events emitted by this command are successfully persisted.
+   */
+  final case class AddItem(
+      itemId: String,
+      quantity: Int,
+      replyTo: ActorRef[StatusReply[Summary]])
+      extends Command
+
+  /**
+   * A command to remove an item from the cart.
+   */
+  final case class RemoveItem(
+      itemId: String,
+      replyTo: ActorRef[StatusReply[Summary]])
+      extends Command
+
+  /**
+   * A command to adjust the quantity of an item in the cart.
+   */
+  final case class AdjustItemQuantity(
+      itemId: String,
+      quantity: Int,
+      replyTo: ActorRef[StatusReply[Summary]])
+      extends Command
+
+  /**
+   * A command to checkout the shopping cart.
+   */
+  final case class Checkout(replyTo: ActorRef[StatusReply[Summary]])
+      extends Command
+
+  /**
+   * A command to get the current state of the shopping cart.
+   */
+  final case class Get(replyTo: ActorRef[Summary]) extends Command
+
+  /**
+   * Summary of the shopping cart state, used in reply messages.
+   */
+  final case class Summary(items: Map[String, Int], checkedOut: Boolean)
+      extends CborSerializable
+
+  /**
+   * This interface defines all the events that the ShoppingCart supports.
+   */
+  sealed trait Event extends CborSerializable {
+    def cartId: String
+  }
+
+  final case class ItemAdded(cartId: String, itemId: String, quantity: Int)
+      extends Event
+
+  final case class ItemRemoved(cartId: String, itemId: String, oldQuantity: Int)
+      extends Event
+
+  final case class ItemQuantityAdjusted(
+      cartId: String,
+      itemId: String,
+      newQuantity: Int,
+      oldQuantity: Int)
+      extends Event
+
+  final case class CheckedOut(cartId: String, eventTime: Instant) extends Event
+
+  def init(implicit system: ActorSystem[_]): Replication[Command] = {
+    val projectionProvider: ReplicationProjectionProvider = R2dbcProjection.atLeastOnceFlow(_, None, _, _)(_)
```

Comment:
This one we'd want a predefined factory method returning ReplicationProjectionProvider for in akka-projection-r2dbc (suggested by Patrik in the benchmark)